Ignore dependencies for and to unsupported sourceSets/variants

jjohannes opened the pull request:
The plugin currently may give wrong advices for source sets that are not yet supported (see #451 and #298).

This PR:
- Adds tests to show the issues.
- Ignores the unsupported cases for now so that users can use these Gradle features and still use this plugin to analyse the dependencies of the source sets that are supported right now.

Problematic cases are:

### Dependencies from unsupported Source Sets/Variants

The plugin produces advices (often wrong) for source sets that are not
yet supported. For example:

```
testFixturesApi("org.apache.commons:commons-collections4:4.4")

extraFeatureApi("org.apache.commons:commons-collections4:4.4")
```

### Dependencies to unsupported Source Sets/Variants

The plugin may produce wrong advices by ignoring the 'requires capabilities' part of dependencies like this:

```
testImplementation(testFixtures(project(':proj')))

api(project(':proj')) { capabilities {
     requireCapabilities('examplegroup:proj-extra-feature')
} }
```

Note: testFixtures(...) also uses 'requires capabilities'

jjohannes commented on src/functionalTest/groovy/com/autonomousapps/jvm/projects/IntegrationTestProject.groovy:
```diff
@@ -133,7 +134,7 @@ final class IntegrationTestProject extends AbstractProject {
     emptyCompAdviceFor(':core'),
     new ComprehensiveAdvice(':proj', [
-      Advice.ofChange(dependency(project('integrationTestImplementation', ':core')), 'api')
+      Advice.ofAdd(transitiveDependency(dependency(':core'), []), 'api')
```
This expectation was "wrong" before. The 'proj.integrationTest' sources do depend on 'core'. So the dependency should not be moved from there.

Unless you assume that 'integrationTest' always inherits all 'implementation' dependencies and the rule is that you do not need to redeclare them then. But this may not be the case (it is not in this test case). There is no guarantee that the setup is like this and thus such assumptions should not be made.

If #451 is implemented, further thoughts could be put into this.

jjohannes commented on src/main/kotlin/com/autonomousapps/internal/utils/gradleStrings.kt:
```diff
@@ -141,50 +141,46 @@ internal fun ComponentIdentifier.resolvedVersion(): String? = when (this) {
  * [ComponentIdentifier.toIdentifier].
  */
-internal fun DependencySet.toIdentifiers(
-  metadataSink: MutableMap<String, Boolean> = mutableMapOf()
```
I think the idea with the metadata can not work like it was here. You would need the metadata for each dependency. E.g. if you have both:

```
api("org.apache.commons:commons-collections4:4.4")
testFixturesApi("org.apache.commons:commons-collections4:4.4")
```

The two declaration would have different metadata. But the approach here only stores one "Metadata" for `org.apache.commons:commons-collections4:4.4`.

It also feels to me like Gradle's dependency resolution should be used here to give you the result (i.e. to which Jar on the other side will the dependency resolve).
If you start storing attributes/capabilities here yourself and pass them around you will somehow end up re-implementing Gradle's variant matching to cover all the cases.

jjohannes replied:
I think, if you want to get to a working solution here later, you should investigate how to do this:

> It also feels to me like Gradle's dependency resolution should be used here to give you the result (i.e. to which Jar on the other side will the dependency resolve).

It's not the goal of this PR to solve this. But feel free to ping me if you get to it later. Happy to discuss/explore this.

autonomousapps replied:
I think one of the problems was doing this in the context of _declarations_, right? At that point, I don't want to _resolve_ any dependencies at all. I just want to know what the user _declared_.

jjohannes commented on src/main/kotlin/com/autonomousapps/internal/utils/gradleStrings.kt:
```diff
@@ -141,50 +141,46 @@ internal fun ComponentIdentifier.resolvedVersion(): String? = when (this) {
-  metadataSink: MutableMap<String, Boolean> = mutableMapOf()
-): Set<String> = mapNotNullToSet {
-  it.toIdentifier(metadataSink)
+internal fun DependencySet.toIdentifiers(): Set<Pair<String, Boolean>> = mapNotNullToSet {
```
The `Pair<String, Boolean>` now stores the information if the dependency is supported or should be ignored for now.

In the future that could then be replaced by a Type that holds more information (metadata of the dependency). For example, the Jar the dependency will resolve to (see comment above).

jjohannes commented on src/main/kotlin/com/autonomousapps/tasks/FindDeclarationsTask.kt:
```diff
@@ -85,40 +82,20 @@ abstract class FindDeclarationsTask : DefaultTask() {
-          // this is not ideal, but it will solve some problems till we can support androidTest analysis.
-          // will match, e.g., "androidTestImplementation", "debugAndroidTestImplementation", and "kaptAndroidTest".
-          it.name.contains("androidTest", true)
-        }
```
This is no longer needed, because it is covered by the more general "supported source sets" implementation.

jjohannes commented on src/main/kotlin/com/autonomousapps/internal/utils/gradleStrings.kt:
```diff
@@ -206,11 +202,21 @@ internal fun Dependency.resolvedVersion(): String? = when (this) {
 }?.intern()
 
-private fun Dependency.isJavaPlatform(): Boolean = when (this) {
-  is ProjectDependency -> dependencyProject.pluginManager.hasPlugin("java-platform")
```
This is not needed IMO. The **ProjectDependency** is also a **ModuleDependency** and it will have the attribute added by `testFixtures(...)` which is checked by the code below.

jjohannes commented on src/test/groovy/com/autonomousapps/model/declaration/VariantSpec.groovy:
```diff
@@ -6,25 +6,28 @@ class VariantSpec extends Specification {
     'implementation'                   | new Variant('main', SourceSetKind.MAIN)
     'debugImplementation'              | new Variant('debug', SourceSetKind.MAIN)
-    'releaseFlavorImplementation'      | new Variant('releaseFlavor', SourceSetKind.MAIN)
+    'flavorReleaseImplementation'      | new Variant('flavorRelease', SourceSetKind.MAIN)
```
I checked with the [Android docs](https://developer.android.com/studio/build/dependencies#dependency_configurations) and with an example project. The naming scheme is always:
`$sourceSet$flavor$buildType$scope`

(sourceSet is empty in case of 'main')

IMO some of the test data here was wrong (switching 'flavor' and 'buildType' order). These won't work anymore as they do not exist in a real Android project.

autonomousapps replied:
Seems like I had a copy-paste error.

autonomousapps reviewed (commented):
I _really_ appreciate this. I want to check out the changes locally and study them a bit to ensure I understand what you've done, but my impression is that it's an improvement.

autonomousapps commented on src/main/kotlin/com/autonomousapps/internal/utils/gradleStrings.kt:
```diff
@@ -141,50 +141,46 @@ internal fun ComponentIdentifier.resolvedVersion(): String? = when (this) {
-      }
-      is ConfigurableFileTree -> files.firstOrNull()?.name
+      }?.let { Pair(it, false) }
+      is ConfigurableFileTree -> files.firstOrNull()?.name?.let { Pair(it, false) }
```
I think we lost the `intern()` calls on these two types.

jjohannes replied:
Good catch. Sorry for that. Added them.

autonomousapps commented on src/main/kotlin/com/autonomousapps/internal/utils/gradleStrings.kt:
```diff
@@ -206,11 +202,21 @@ internal fun Dependency.resolvedVersion(): String? = when (this) {
+ * Return true if the dependency does not point at the "main" variant of the other module.
+ * Other variants are for example testFixtures() or Feature Variants with different capabilities.
+ */
+private fun Dependency.isDependencyToNonMainFeature(): Boolean = when (this) {
```
Could you add links to the documentation for test fixtures and feature variants in your kdoc? I have a fairly verbose style when it comes to this documentation because the domain is so complex and it's easy to forget context as time passes.

autonomousapps commented on src/main/kotlin/com/autonomousapps/subplugin/ProjectPlugin.kt:
```diff
@@ -692,6 +692,25 @@ internal class ProjectPlugin(private val project: Project) {
     }
   }
 
+  private fun Project.supportedSourceSetNames() = provider {
```
When I checked the explicit type of this, I got `Provider<Collection<String>>`, when I would have expected `Provider<Set<String>>` (from line 710). I think we can achieve that (which I prefer), with this small change:
```kotlin
the<AppExtension>().applicationVariants.flatMapToSet {
  it.sourceSets.map { sourceSet -> sourceSet.name } +
    (it.unitTestVariant?.sourceSets?.map { sourceSet -> sourceSet.name } ?: emptySet())
  // Not yet supported: + it.testVariant.sourceSets.map { sourceSet -> sourceSet.name }
}
```
Note the use of the `flatMapToSet` extension function.

(the same change works for the `LibraryExtension` section.

autonomousapps replied:
And then as a style matter, could you move this whole function _below_ the one that uses it? Semantically, I see this as more of a helper function, and therefore less important than the `configureAggregationTasks()` function.

jjohannes replied:
Yes that's better. I made the changes and also added a short Javadoc comment to the function.

autonomousapps reviewed (approved):
Especially given that all the tests pass, I'm very comfortable with these changes. I've left a few notes -- would you mind taking a look?

autonomousapps commented:
FYI, I just [merged a PR](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/pull/656) that will conflict with `IntegrationTestProject`. Should be straightforward to resolve, and hopefully you'll agree it has simplified test fixture setup a bit :)

jjohannes commented on src/functionalTest/groovy/com/autonomousapps/jvm/projects/FeatureVariantTestProject.groovy:
```diff
@@ -0,0 +1,113 @@
+    actualBuildHealth(gradleProject)
+  }
+
+  // Note: The 'proj-extra.jar' is considered part of the 'main variant' of ':proj', which is not correct.
```
In general yes. Although I haven't looked in detail, why this particular thing is happening. This is not changed by this PR. The test just reveals the existing behavior.

IMO this particular "false positive" is not super critical. Because often if you use an "extra feature" of a project, you also use the main variant and would have that dependency anyway.

jjohannes commented on src/functionalTest/groovy/com/autonomousapps/jvm/projects/FeatureVariantTestProject.groovy:
```diff
@@ -0,0 +1,113 @@
+  ]
+
+  final List<ComprehensiveAdvice> expectedBuildHealth = [
+    // Not yet implemented: missing advice to move the dependency to 'extra' to implementation
```
No I meant 'to', but the word 'to' is used much too often in that sentence for anyone to understand. 😄 I expanded the comments so that it is (hopefully) clearer what is missing.

jjohannes commented:
Thank you for the detailed review @autonomousapps. I rebased and addressed the feedback. Let me know if there is anything else you would like to have addressed before merging this.
